src/zelos/ext/plugins/snapshot/snapshot.py

```diff
@@ -0,0 +1,205 @@
+        for region in sorted(regions):
+            addr = region.address
+            size = region.size
+            perm = region.prot
+            name = "<unk>"
+            kind = "<unk>"
+            if self.memory.get_region(addr) is not None:
+                region = self.memory.get_region(addr)
+                name = region.name
+                kind = region.kind
+            if addr == 0x80000000:
+                continue  # GDT only
+            dumped = False
+
+            # Dump main binary
+            if kind == "main" or name == "main":
+                if "base_address" not in out_map:
+                    out_map["base_address"] = addr
+                section = {}
+                section_name = name
+                tmpname = name.split(" ")
+                if len(tmpname) > 1:
+                    section_name = tmpname[1]
+
+                section["name"] = section_name
+                section["address"] = addr
+                section["permissions"] = perm
+                # Temporary hack. The MEW packer requires executable
+                # header section. But, we mark it non-executable for the
+                # dump.
+                if section_name == ".pe":
+                    section["permissions"] = 0x1
+                data = self.memory.read(addr, size)
+                section["data"] = base64.b64encode(data).decode()
+                dumped = True
+
+            # Dump main and thread stacks binary
+            if kind == "stack" and "dll_main" not in name:
+                if "base_address" not in out_map:
+                    out_map["base_address"] = addr
+                section = {}
+                section["name"] = "stack_" + name
+                section["address"] = addr
+                section["permissions"] = perm
+                data = self.memory.read(addr, size)
+                section["data"] = base64.b64encode(data).decode()
+                dumped = True
+
+            # Dump heap, sections, virtualalloc'd regions. Note that currently
+            # we don't make use of dynamically allocated heaps, and so they are
+            # excluded. Once that changes, we should include them here
+            if (
+                (kind == "heap" and name != "heap")
+                or kind == "valloc"
+                or kind == "section"
+            ):
+                if "base_address" not in out_map:
+                    out_map["base_address"] = addr
+                section = {}
+                section["name"] = kind + "_" + name
+                section["address"] = addr
+                section["permissions"] = perm
+                data = self.memory.read(addr, size)
+                if kind == "heap" and name == "main_heap":
+                    # Truncate unused portion of heap
+                    section["data"] = base64.b64encode(
+                        data[
+                            : self.memory.heap.current_offset
+                            - self.memory.HEAP_BASE
```

Comment:
This memory should be accessed from internal_engine rather than self to distinguish it from the Memory API